### PR TITLE
[BREAKING] Enforce STATICCALL for view and pure

### LINKED ISSUE #4089: Always use staticcall for pure (not only in 0.5.0 mode)
When we introduced to use `STATICCALL` for `view` and `pure` we enabled that for 0.5.0 only, in order not to break non-compliant `view`/`constant` functions.

However we never made such a commitment for `pure`, since it was newly introduced, so we could have enabled `staticcall` for that unconditionally.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #4089.

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1810,14 +1810,12 @@ void ExpressionCompiler::appendExternalFunctionCall(
 	if (_functionType.bound())
 		utils().moveToStackTop(gasValueSize, _functionType.selfType()->sizeOnStack());
 
-	bool const v050 = m_context.experimentalFeatureActive(ExperimentalFeature::V050);
 	auto funKind = _functionType.kind();
 	bool returnSuccessCondition = funKind == FunctionType::Kind::BareCall || funKind == FunctionType::Kind::BareCallCode || funKind == FunctionType::Kind::BareDelegateCall;
 	bool isCallCode = funKind == FunctionType::Kind::BareCallCode || funKind == FunctionType::Kind::CallCode;
 	bool isDelegateCall = funKind == FunctionType::Kind::BareDelegateCall || funKind == FunctionType::Kind::DelegateCall;
 	bool useStaticCall =
```
Comment (erak) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Could this now be one line?

**Review by erak — DISMISSED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Yep :)

**Review by ekpyron — COMMENTED** [review]
Changelog entry :D.

**Comment by ekpyron** [comment]
And I wonder whether there should be more test cases for this...

**Comment by erak** [comment]
@ekpyron I also thought about that. Would you recommend to test that ``payable`` and ``nonpayable`` do not use ``STATICCALL``, also in combination with different EVM versions?

**Comment by ekpyron** [comment]
@bit-shift I thought about adding more end to end tests that expect reverts on state changes when staticcall is used... but there's one test for that, maybe that suffices...
The other way round is probably rather well tested - every non-reverting (external) function call in the end-to-end tests that changes state is a test for that. I suspect there are a few of those, but I didn't check :).
So I'm not sure - it's probably fine as it is already...

**Comment by ekpyron** [comment]
@bit-shift Maybe you're right - I just quickly looked through the end-to-end tests and it's fewer storage-writing external calls in there than I'd have expected that test this (maybe I missed a lot of them, though)...

So maybe it does make sense to add an explicit tests to ensure that payable and non-payable calls don't staticcall and don't revert when writing to the storage, just to be sure.

**Comment by leonardoalt** [comment]
Added a test for that

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -12430,6 +12430,43 @@ BOOST_AUTO_TEST_CASE(senders_balance)
 	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(27)));
 }
 
+BOOST_AUTO_TEST_CASE(write_storage)
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp]: ``write_storage_external`` maybe?

**Review by ekpyron — DISMISSED** [review]
Apart from maybe changing the name of the new test, it looks fine to me.

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Yea, better

**Review by ekpyron — CHANGES_REQUESTED** [review]
I just noticed: ``docs/contracts.rst`` has a section regarding ``View Functions`` that refers to STATICCALL and needs to be updated.

**Comment by leonardoalt** [comment]
Thanks, missed that.
Updated now.

**Review by ekpyron — DISMISSED** [review]

File: Changelog.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ Breaking Changes:
  * ABI Encoder: Properly pad data from calldata (``msg.data`` and external function parameters). Use ``abi.encodePacked`` for unpadded encoding.
  * Code Generator: Signed right shift uses proper arithmetic shift, i.e. rounding towards negative infinity. Warning: this may silently change the semantics of existing code!
  * Code Generator: Revert at runtime if calldata is too short or points out of bounds. This is done inside the ``ABI decoder`` and therefore also applies to ``abi.decode()``.
+ * Code Generator: Use STATICCALL for pure and view functions. This was already the case in the experimental 0.5.0 mode.
```
Comment (erak) [inline Changelog.md]: Should this be preformatted? "Use STATICCALL for ``pure`` and ``view`` functions."

**Review by erak — DISMISSED** [review]
Besides the tiny comment about formatting the Changelog, it looks good to me.

**Review by ekpyron — DISMISSED** [review]

**Comment by leonardoalt** [comment]
Improved doc, rebased and squashed

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
This reminds me: do we have a test case which tries to do state modification through inline assembly (and is not detected), but it doesn't actually changes the state due to staticcall?

I guess we do detect all these possibilities in inline assembly.

**Comment by chriseth** [comment]
We have tests to detect state modification through inline assembly and thus change the state mutability of the function, so I think a semantics test would be redundant, but we can still add it of course.

**Pull request merged**